internal/sync/lifecycle_test.go

```diff
@@ -415,3 +439,24 @@ func TestLifecycleOnce(t *testing.T) {
 		})
 	}
 }
+
+// TestStopping verifies that a lifecycle object can spawn a goroutine and wait
+// for that goroutine to exit in its stopping state.  The goroutine must wrap
+// up its work when it detects that the lifecycle has begun stopping.  If it
+// waited for the stopped channel, the stop callback would deadlock.
+func TestStopping(t *testing.T) {
+	l := Once()
+	l.Start(nil)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		<-l.Stopping()
+		wg.Done()
+	}()
+
+	l.Stop(func() error {
+		wg.Wait()
```

Comment:
Instead of deadlocking (and waiting for test timeout, maybe have something that uses `time.After` and have a test timeout of 5 seconds?